### PR TITLE
[MRG] BUG: fix elusive MPI runtime error in tests

### LINKED ISSUE #406: MPI test fails locally
I've run across this in current `master` when running `test_parallel_backends.py:test_run_mpibackend_oversubscribed`, can any other macOS users confirm?

```bash
E           RuntimeError: MPI simulation failed. Return code: 143

hnn_core/parallel_backends.py:235: RuntimeError
----------------------------- Captured stdout call -----------------------------
MPI will run over 12 processes
Running 2 trials...
numprocs=12
```

This is running on my 4-core Air, so I'd expect `1.5 * 4 == 6` cores, not 12. My environment is:

```
mpi                       1.0                     openmpi    conda-forge
mpi4py                    3.0.3            py38h368b8e5_5    conda-forge
openmpi                   4.1.0                h5807002_1    conda-forge
neuron                    7.8.2                    pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by rythorpe**
closes #406 

Finally figured it out. Our `MPIBackend` unit tests use a reduced network model to speed things up. This revealed a rare edge case that only shows up on machines with a large number of logical cores, where more MPI threads were created than there were neurons contributing to the dipole. When `NetworkBuilder().aggregate_data()` then tried to add up and reduce its NEURON data instances across both intra-rank and inter-rank neurons, it hung due to to a mismatch between `h.Vector()` sizes.

### CODE REVIEW

File: hnn_core/parallel_backends.py [hunk 1]
```diff
@@ -705,7 +705,13 @@ def simulate(self, net, tstop, dt, n_trials, postproc=False):
                                                     postproc=postproc)
 
         print("Running %d trials..." % (n_trials))
-        dpls = []
```
Comment (rythorpe) [inline hnn_core/parallel_backends.py:708]: Don't know why this was here, so I cut it.

File: hnn_core/parallel_backends.py [hunk 2]
```diff
@@ -705,7 +705,13 @@ def simulate(self, net, tstop, dt, n_trials, postproc=False):
...
+
+        if self.n_procs > net._n_cells:
+            raise ValueError(f'More MPI processes were assigned than there '
+                             f'are cells in the network. Please decrease '
+                             f'the number of parallel processes (got n_procs='
+                             f'{self.n_procs}) over which you will '
+                             f'distribute the {net._n_cells} network neurons.')
```
Comment (rythorpe) [inline hnn_core/parallel_backends.py:714]: This also prevents a rare edge case where some MPI processes might get created without being assigned any NEURON cells instances.

**Comment by rythorpe** [comment]
@cjayb All the tests on your M1 Mac should pass after this PR. Feel free to give it a try if you feel inspired!

**Comment by cjayb** [comment]
> @cjayb All the tests on your M1 Mac should pass after this PR. Feel free to give it a try if you feel inspired!

I did, and they do! Amazing sleuthing @rythorpe —tip of the old hat to you, Sir!

**Comment by jasmainak** [comment]
Very cool! What was the reason that the test was failing intermittently? Is their some non-deterministic part in the rank assignment?

**Comment by cjayb** [comment]
> What was the reason that the test was failing intermittently?

I'd call this a consistent issue for those of us with more >6 logical cores. The oversubscription test creates `1.5x` threads—for my 8 cores this is 12. There are `3 x 3 == 9` grid points in the test network, each with 2 pyramidal cells that contribute to the aggregate dipole moment. The `gid` assignment starts by looping over cells, which 'run out' at 18—that leaves `24-18==6` threads without a cell that contributes, which chokes the post-simulation reduction operations.

Sound right @rythorpe ?

The CI tests run on less than 6 cores, hence no problem.

**Comment by rythorpe** [comment]
Yes, that's exactly it!

File: doc/whats_new.rst [hunk 3]
```diff
@@ -54,6 +54,11 @@ Bug
 - Fix bug where :func:`~hnn_core.network.pick_connection` did not return an
   empty list when searching non existing connections, by `Nick Tolley`_ in :gh:`515`
 
+- Fix bug in :meth:`~hnn_core.NetworkBuilder.aggregate_data` where continuous
```
Comment (jasmainak) [inline doc/whats_new.rst:57]: should we just say that it created intermittent failures with MPIBackend without mention of `NetworkBuilder` object? I think that's mostly for internal use, no?

File: hnn_core/network_builder.py [hunk 4]
```diff
@@ -530,25 +530,45 @@ def _record_spikes(self):
...
-                else:
-                    nrn_dpl.append(cell.dipole)
+                nrn_dpl.add(cell.dipole)
 
             self._vsoma[cell.gid] = cell.rec_v
             self._isoma[cell.gid] = cell.rec_i
 
-        _PC.allreduce(self._nrn_dipoles['L5_pyramidal'], 1)
-        _PC.allreduce(self._nrn_dipoles['L2_pyramidal'], 1)
+        # reduce across threads
```
[the remark this answers is not in the document]
Comment (jasmainak) [reply hnn_core/network_builder.py:569]: oh, or maybe it is threads? when they are oversubscribed? not sure what's the right terminology in MPI parlance

[the remark this answers is not in the document]
Comment (rythorpe) [reply hnn_core/network_builder.py:569]: Yeah, it’s not necessarily cores due to oversubscribing and hyperthreading. I think “threads” or MPI “processes” are both correct, though I’m far from an authority on the topic 🤷

File: doc/whats_new.rst (hunk 3 above)
Comment (rythorpe) [reply doc/whats_new.rst:57]: I feel like what I put is more explicit, but I’ll defer to your judgment re: the appropriate scope for whats_new.rst (it’s still a little hazy in my head how anyone but an advanced user might take advantage of the documentation in this file).

File: hnn_core/network_builder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasmainak) [reply hnn_core/network_builder.py:569]: yeah "threads" might be more appropriate now that I think more about it

File: doc/whats_new.rst (hunk 3 above)
Comment (jasmainak) [reply doc/whats_new.rst:57]: This document is supposed to be user facing ... e.g., when you install a new release, what are the things you need to be aware of as a user. We sometimes go into the details too much. Another purpose is to give folks credit for their work

Comment (rythorpe) [reply doc/whats_new.rst]: Okay, what about this: https://github.com/jonescompneurolab/hnn-core/pull/545/commits/3528e69686fc7f17e5bbebf0a79c64c937ce3d20?

Comment (rythorpe) [reply doc/whats_new.rst]: Maybe still too much detail? :grimacing:

**Comment by jasmainak** [comment]
looks good! I'm trying to enable auto-merge ... will see if I can make it work on this PR

**Comment by jasmainak** [comment]
Thanks @rythorpe !!! 🥳 This certainly deserves a celebration
